src/Stream.ts

```diff
@@ -0,0 +1,108 @@
+import { Effect } from "./Effect";
+import * as Fx from "./Fx";
+import * as Task from "./Task";
+import * as Async from "./Async";
+import * as Fail from "./Fail";
+import * as Fork from "./Fork";
+
+export class Stream<A> extends Effect('Stream')<A, void> { }
+
+export namespace Stream { 
+  export type Event<T> = T extends Stream<infer A> ? A : never
+}
+
+export function event<const A>(a: A): Fx.Fx<Stream<A>, void> { 
+  return new Stream(a)
+}
+
+export function observe<E, R, E2>(fx: Fx.Fx<E, R>, f: (a: Stream.Event<E>) => Fx.Fx<E2, void>): Fx.Fx<Exclude<E, Stream<any>> | E2, R> { 
+  return fx.pipe(Fx.handle<typeof Stream, E2>(Stream, (a) => f(a as Stream.Event<E>)))
+} 
+
+export function collectAll<E, A>(fx: Fx.Fx<E, A>): Fx.Fx<Exclude<E, Stream<any>>, readonly Stream.Event<E>[]> {
+  return Fx.fx(function* () {
+    const events: Stream.Event<E>[] = []
+    yield* observe(fx, a => { 
+      events.push(a)
+      return Fx.unit
+    })
+    return events
+  })
+}
+
+export function map<E, A, B>(fx: Fx.Fx<E, A>, f: (a: Stream.Event<E>) => B): Fx.Fx<Exclude<E, Stream<any>> | Stream<B>, A> { 
+  return observe(fx, a => event(f(a)))
+}
+
+export function filter<E, A>(fx: Fx.Fx<E, A>, f: (a: Stream.Event<E>) => boolean): Fx.Fx<Exclude<E, Stream<any>> | Stream<Stream.Event<E>>, A> {
+  return observe(fx, a => f(a) ? event(a) : Fx.unit)
+}
+
+export function flatMap<E, X, B, E2>(fx: Fx.Fx<E, X>, f: (a: Stream.Event<E>) => Fx.Fx<E2, B>): Fx.Fx<Fork.Fork | Async.Async | E | E2 | Stream<B>, X> {
+  return Fx.bracket(
+    Fx.sync(() => new TaskSet<E | E2>()),
+    (tasks) => Fx.sync(() => tasks[Symbol.dispose]()),
+    (set) => Fx.fx(function* () {
+      const x = yield* observe(fx, a => set.run(f(a)))
+      yield* set.wait()
```

Comment:
I'm a little surprised this works.  I wonder if it's possible for `observe` to end up adding things to `set` after line 47 has executed.  Can that happen?

Reply:
I don't think so. When the Fx returns its also when the handler `observe` adds is removed from the stack. Is it at all possible for inner fibers to continue using them?

In my mind the return _should_ signal the end of the Stream and no further events, which is why at that point its just necessary to wait on any fibers left.

Writing that last bit made me remember I probably need to remove tasks from the set when they complete

Reply:
We might need more testing to figure it out.  I saw you dropped flatMap in the latest, which seems like the right thing for now.  Thanks!